### PR TITLE
add info about implementing store.js methods

### LINKED ISSUE #349: update Readme to include details on extending Store implementation from store.js
As noted in a comment on #97, developers who want to write a Store implementation either need to:

1. extend their Store from express-session's store.js, which typically means requiring express-session as a dependency
2. implement their own version of the functionality in store.js, session.js, and cookie.js (or copy the files from express-session)

This is because express-session requires the following methods on the Store in order to work:
* regenerate()
* load()
* createSession()

These are available in store.js, but `createSession` itself requires both session.js and cookie.js.

The documentation should be updated to provide guidance on these facts.

### PR DESCRIPTION
**Pull request by dwhieb**
closes #349

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,6 +54,8 @@ Specifies the value for the `Domain` `Set-Cookie` attribute. By default, no doma
 is set, and most clients will consider the cookie to apply to only the current
 domain.
 
+**Note:** Some browsers will not set a cookie on `localhost` is the `domain` attribute is set, so you should omit this setting when testing on `localhost`.
```
Comment (dougwilson) [inline README.md]: Please make a separate PR for this note, as it is not related to "info about implementing store.js methods"

File: README.md [hunk 2]
```diff
@@ -577,7 +628,7 @@ and other multi-core embedded devices).
 [connect-sqlite3-url]: https://www.npmjs.com/package/connect-sqlite3
 [connect-sqlite3-image]: https://img.shields.io/github/stars/rawberg/connect-sqlite3.svg?label=%E2%98%85
 
-[![★][documentdb-session-image] documentdb-session][documentdb-session-url] A session store for Microsoft Azure's [DocumentDB](https://azure.microsoft.com/en-us/services/documentdb/) NoSQL database service.
+[![★][documentdb-session-image] documentdb-session][documentdb-session-url] A session store for Microsoft Azure's  [DocumentDB](https://azure.microsoft.com/en-us/services/documentdb/) NoSQL database service.
```
Comment (dougwilson) [inline README.md]: Looks like you accidentally included adding a stray space in the PR

**Comment by dougwilson** [comment]
Can you add newlines to your paragraphs to keep the maximum line length down, similar to the other parts of the doc?

File: README.md [hunk 3]
```diff
@@ -410,6 +412,37 @@ and **optional**.
   * Optional methods are ones this module does not call at all, but helps
     present uniform stores to users.
 
+Some methods (`createSession`, `load`, and `regenerate`) are already available from the `express-session` module. These methods are marked as "available from `express-session`" below. Instead of writing these methods yourself, it is recommended that you pass the `express-session` object to your store on initialization, and extend your store from `expressSession.Store`. See below for an example:
+
+In `app.js/server.js/index.js/etc.`:
+
+```
```
Comment (dougwilson) [inline README.md]: Mark this as a JavaScript code block.

File: README.md [hunk 4]
```diff
@@ -410,6 +412,37 @@ and **optional**.
...
+const MyStore = require('my-store')(expressSession);
+
+app.use(expressSession({
+  store: new MyStore(options),
+  secret: 'mycookiesecret'
+}));
+```
+
+In your store's `index.js` (main) file:
+```
```
Comment (dougwilson) [inline README.md]: Mark this as a JavaScript code block.

File: README.md [hunk 5]
```diff
@@ -434,6 +467,12 @@ the session is destroyed.
 This optional method is used to delete all sessions from the store. The
 `callback` should be called as `callback(error)` once the store is cleared.
 
+### store.createSession(req, session)
+
+**Required** (available from `express-session`)
+
+This required method is used to create a session from JSON `session` data, set `session.cookie`, and set `req.session`. It should return the new session object. This method will need to utilize `express-session`'s `Cookie` and `Session` objects.
```
Comment (dougwilson) [inline README.md]: Does it _need_ to, or would using it simply make it easier to implement? I don't think we do instanceof checks anywhere that makes using them actually a requeirment.

File: README.md [hunk 6]
```diff
@@ -452,6 +491,18 @@ The `session` argument should be a session if found, otherwise `null` or
 `undefined` if the session was not found (and there was no error). A special
 case is made when `error.code === 'ENOENT'` to act like `callback(null, null)`.
 
+### store.load(sid, callback)
+
+**Required** (available from `express-session`)
+
+This required method is used to load a `Session` instance via the given session ID (`sid`). The `callback` should be called as `callback(error, session)`. This method calls `store.createSession` internally.
```
Comment (dougwilson) [inline README.md]: Can you clarify what "This method calls store.createSession internally" means in the docs? Reading the docs, it sounds like I can either make it, or using the one from the module. The rest of the doc is about writing it myself, and so does that mean I have to call that method internally if I write it myself?

File: README.md [hunk 7]
```diff
@@ -452,6 +491,18 @@ The `session` argument should be a session if found, otherwise `null` or
...
+
+### store.regenerate(req, callback)
+
+**Required** (available from `express-session`)
+
+This required method is used to regenerate the given request's (`req`) session. The `callback` should be called as `callback(error)`.
```
Comment (dougwilson) [inline README.md]: Can you describe in more detail what "regenerate" means? How would I go about writing this method if I'm not using the one from the module?

**Comment by dougwilson** [comment]
I left a bunch of comments on fixes. Thank you so much for taking the time to contribute this! I'll add you to the package.json contributors list when merging if that is agreeable, as this will be an awesome contribution that's been waiting a long time 👍

File: README.md (hunk 1 above)
Comment (dwhieb) [inline README.md]: b962203fafd8a8c236ea19d5c4b1b4920602774a

Comment (dwhieb) [inline README.md]: https://github.com/expressjs/session/pull/354/commits/b962203fafd8a8c236ea19d5c4b1b4920602774a

File: README.md (hunk 3 above)
Comment (dwhieb) [inline README.md]: Addressed in https://github.com/expressjs/session/pull/354/commits/9c0ea61fb2f842d54cbbbb535d2a0b5a9bae98f2

File: README.md (hunk 4 above)
Comment (dwhieb) [inline README.md]: Addressed in https://github.com/expressjs/session/pull/354/commits/9c0ea61fb2f842d54cbbbb535d2a0b5a9bae98f2

File: README.md (hunk 5 above)
Comment (dwhieb) [inline README.md]: Addressed in https://github.com/expressjs/session/pull/354/commits/2811daa8d74d0a20faeb940a3fb08354c19257a0

File: README.md (hunk 6 above)
Comment (dwhieb) [inline README.md]: Addessed in https://github.com/expressjs/session/pull/354/commits/5fbf848ba15048fb4f36ea525a6e086ffc62b186

**Comment by dwhieb** [comment]
@dougwilson Ok it should be all set! I've added line breaks for readability wherever I could. Each of your other comments were addressed by a particular commit - I've tried to make the commit messages transparent, so you can see which commit goes with which comment. I also replied to each of your comments with the relevant commit number. Let me know if there's anything else I can do.

I'd certainly be okay with being added to the contributors list. This is a great package that I've come to depend on, so I was happy to help with it.

**Comment by dougwilson** [comment]
Sweet! So one thing is do we really need to add links into the source code? Right now all those links are pointing to specific line numbers against whatever just happens to be on the `master` branch. It seems like it is very likely those links will constantly get out-of-sync.

**Comment by dougwilson** [comment]
@dwhieb thoughts on the above?

**Comment by dougwilson** [comment]
I'm going to close this since I haven't heard anything back.

**Comment by demurgos** [comment]
Any news about this PR?
